File: client-apps/cli/embedded/agentrunner/source/src/stigmer_runner/worker/activities/graphton/writeback_coordinator.py
"""Incremental git write-back coordinator for workspace entries.

During agent execution, each file-modifying tool call (write, edit)
triggers an incremental commit-and-push cycle for the affected git
workspace.  The first cycle creates the branch and PR; subsequent
cycles add commits to the same branch — the PR updates automatically
on GitHub.

This replaces the batch ``writeback.py`` approach where all git
operations happened after the LangGraph stream ended.  The user now
sees the PR link the moment the first file is written, and watches
the diff grow in real time.

Lifecycle:
    1.  Created after workspace provisioning in ``execute_graphton``.
    2.  ``on_file_modified(path)`` called from ``StreamExecutor`` on
        each ``on_tool_end`` for file-modifying tools.
    3.  ``finalize()`` called from ``post_stream`` as a safety net.

Concurrency: one ``asyncio.Lock`` per workspace entry serializes
git operations.  Background tasks spawned by ``StreamExecutor``
are tracked in ``pending_tasks`` so ``post_stream`` can drain them
before calling ``finalize()``.
"""

from __future__ import annotations

import asyncio
import dataclasses
import logging
from typing import TYPE_CHECKING, Any

import httpx
from ai.stigmer.agentic.agentexecution.v1.writeback_pb2 import (
    WorkspaceWriteBack,
    WorkspaceWriteBackPhase,
)
from ai.stigmer.agentic.session.v1.enum_pb2 import (
    GIT_WRITE_BACK_BRANCH_AND_PR,
    GIT_WRITE_BACK_MODE_UNSPECIFIED,
)
from graphton.core.github_api import (
    github_create_pr,
    parse_github_repo,
    parse_token_from_credentials,
)

from stigmer_runner.worker.workspace.backend import ExecuteResult

# Write-back modes treated as "enabled" by the platform.
# UNSPECIFIED = platform decides (currently: enabled when credentials exist).
# BRANCH_AND_PR = explicit opt-in.
# Any future opt-out enum (e.g. DISABLED = 2) is excluded automatically.
_WRITE_BACK_ENABLED_MODES: frozenset[int] = frozenset({
    GIT_WRITE_BACK_MODE_UNSPECIFIED,
    GIT_WRITE_BACK_BRANCH_AND_PR,
})

if TYPE_CHECKING:
    from stigmer_runner.worker.activities.graphton.status_builder import StatusBuilder
    from stigmer_runner.worker.workspace import ProvisionResult

_CREDENTIAL_FILE = "/home/daytona/.git-credentials"


@dataclasses.dataclass
class _EntryState:
    """Mutable per-workspace state tracked across incremental cycles."""

    branch_created: bool = False
    pr_created: bool = False
    pr_url: str = ""
    pr_number: int = 0
    commit_count: int = 0
    last_commit_sha: str = ""
    github_token: str = ""
    github_owner: str = ""
    github_repo: str = ""


@dataclasses.dataclass(frozen=True)
class _EligibleEntry:
    """Immutable descriptor for a workspace entry eligible for write-back."""

    provision_result: Any  # ProvisionResult
    base_branch: str
    root_dir: str
    entry_name: str


class WriteBackCoordinator:
    """Manages incremental git write-back for workspace entries during execution.

    Thread-safe (via asyncio locks) and idempotent: calling
    ``on_file_modified`` when there are no uncommitted changes is a no-op.
    """

    def __init__(
        self,
        *,
        status_builder: StatusBuilder,
        execution_id: str,
        provision_results: list[ProvisionResult],
        workspace_entries: list[Any],
        sandbox: Any,
        workspace_backend: Any,
        logger: logging.Logger,
    ) -> None:
        self._sb = status_builder
        self._execution_id = execution_id
        self._sandbox = sandbox
        self._workspace_backend = workspace_backend
        self._log = logger
        self._short_id = execution_id[:8]
        self._branch_name = f"stigmer/{self._short_id}"

        self._eligible: dict[str, _EligibleEntry] = {}
        self._state: dict[str, _EntryState] = {}
        self._locks: dict[str, asyncio.Lock] = {}

        self._pending: set[asyncio.Task[None]] = set()

        self._init_eligible_entries(provision_results, workspace_entries)

    @property
    def has_eligible_entries(self) -> bool:
        """True when at least one workspace entry is configured for write-back."""
        return len(self._eligible) > 0

    @property
    def pending_tasks(self) -> set[asyncio.Task[None]]:
        """Background write-back tasks not yet completed."""
        self._pending = {t for t in self._pending if not t.done()}
        return set(self._pending)

    def track_task(self, task: asyncio.Task[None]) -> None:
        """Register a background task for later draining."""
        self._pending.add(task)
        task.add_done_callback(self._pending.discard)

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    async def on_file_modified(self, path: str) -> None:
        """Called after a file-modifying tool completes.

        Resolves *path* to a workspace entry and runs an incremental
        commit/push cycle.  No-op if the path is not in an eligible
        workspace or there are no uncommitted changes.
        """
        entry_name = self._resolve_entry(path)
        if entry_name is None:
            return

        async with self._locks[entry_name]:
            await self._incremental_write_back(entry_name)

    async def finalize(self) -> None:
        """Post-execution safety net.

        Checks every eligible workspace entry for remaining uncommitted
        changes and commits/pushes them.  Catches files modified by
        shell commands or other paths not triggered by ``on_file_modified``.
        """
        for entry_name in self._eligible:
            try:
                async with self._locks[entry_name]:
                    await self._incremental_write_back(entry_name)
            except Exception as exc:
                self._log.warning(
                    "[WRITE_BACK] execution=%s entry=%s — finalize error: %s",
                    self._execution_id, entry_name, exc,
                )

    # ------------------------------------------------------------------
    # Initialization
    # ------------------------------------------------------------------

    def _init_eligible_entries(
        self,
        provision_results: list[Any],
        workspace_entries: list[Any],
    ) -> None:
        """Build the lookup of eligible workspace entries."""
        mode_map: dict[str, int] = {}
        for entry in workspace_entries:
            source = entry.source
            if source.HasField("git_repo"):
                mode_map[entry.name] = source.git_repo.write_back_mode

        for pr in provision_results:
            if pr.source_type.value != "git_repo":
                continue
            if pr.git_metadata is None:
                continue
            if not pr.git_metadata.git_credentials_configured:
                continue
            if mode_map.get(pr.entry_name, 0) not in _WRITE_BACK_ENABLED_MODES:
                continue

            self._eligible[pr.entry_name] = _EligibleEntry(
                provision_result=pr,
                base_branch=pr.git_metadata.branch,
                root_dir=pr.root_dir,
                entry_name=pr.entry_name,
            )
            self._state[pr.entry_name] = _EntryState()
            self._locks[pr.entry_name] = asyncio.Lock()

        if self._eligible:
            self._log.info(
                "[WRITE_BACK] execution=%s — coordinator initialized with "
                "%d eligible workspace(s): %s",
                self._execution_id,
                len(self._eligible),
                list(self._eligible.keys()),
            )

    # ------------------------------------------------------------------
    # Path resolution
    # ------------------------------------------------------------------

    def _resolve_entry(self, path: str) -> str | None:
        """Map a tool-call path to the owning workspace entry name.

        For single-entry sessions the tool path is relative to the
        backend root which equals the entry's ``root_dir``.  For
        multi-entry sessions each entry sits under
        ``{backend_root}/{entry_name}/``, so the path prefix determines
        ownership.

        Returns ``None`` when the path does not fall inside any eligible
        workspace or when there is only one eligible entry (skip the
        resolution and return it directly).
        """
        if not self._eligible:
            return None

        if len(self._eligible) == 1:
            return next(iter(self._eligible))

        normalized = path.lstrip("/")
        for entry_name in self._eligible:
            prefix = entry_name + "/"
            if normalized.startswith(prefix) or normalized == entry_name:
                return entry_name

        return None

    # ------------------------------------------------------------------
    # Core incremental write-back
    # ------------------------------------------------------------------

    async def _incremental_write_back(self, entry_name: str) -> None:
        """Run one commit/push cycle for a single workspace entry.

        Creates branch and PR on first invocation; subsequent calls add
        commits to the existing branch.
        """
        eligible = self._eligible[entry_name]
        state = self._state[entry_name]
        root_dir = eligible.root_dir

        def _exec(cmd: str, timeout: int = 15) -> ExecuteResult:
            full_cmd = f"cd {root_dir} && {cmd}"
            if self._sandbox is not None:
                raw = self._sandbox.process.exec(full_cmd, timeout=timeout)
                return ExecuteResult(
                    exit_code=getattr(raw, "exit_code", 1),
                    stdout=getattr(raw, "stdout", None) or getattr(raw, "output", None) or "",
                    stderr=getattr(raw, "stderr", "") or "",
                )
            return self._workspace_backend.execute(full_cmd, timeout=timeout)

        mutation_started = False
        try:
            if not self._has_changes(_exec):
                return

            mutation_started = True

            if not state.branch_created:
                await self._create_branch(entry_name, state, _exec)

            commit_msg = f"agent changes ({state.commit_count + 1})"
            await self._commit_and_push(entry_name, state, _exec, commit_msg)

            if not state.pr_created:
                await self._create_pr(entry_name, state, eligible, _exec)

            self._update_status(entry_name, state, eligible, _exec)

        except Exception as exc:
            self._log.warning(
                "[WRITE_BACK] execution=%s entry=%s — incremental error: %s",
                self._execution_id, entry_name, exc,
            )
            if not mutation_started:
                return
            wb = WorkspaceWriteBack(
                workspace_entry_name=entry_name,
                base_branch=eligible.base_branch,
                branch_name=self._branch_name if state.branch_created else "",
                phase=WorkspaceWriteBackPhase.WORKSPACE_WRITE_BACK_FAILED,
                error=str(exc),
            )
            if state.pr_created:
                wb.pull_request_url = state.pr_url
                wb.pull_request_number = state.pr_number
            self._sb.add_workspace_write_back(wb)

    # ------------------------------------------------------------------
    # Git operations
    # ------------------------------------------------------------------

    @staticmethod
    def _has_changes(exec_fn: Any) -> bool:
        """Return True when the working tree has uncommitted changes."""
        diff = exec_fn("git diff --stat")
        staged = exec_fn("git diff --cached --stat")

        if (diff.stdout and diff.stdout.strip()) or (
            staged.stdout and staged.stdout.strip()
        ):
            return True

        untracked = exec_fn("git ls-files --others --exclude-standard")
        return bool(untracked.stdout and untracked.stdout.strip())

    async def _create_branch(
        self,
        entry_name: str,
        state: _EntryState,
        exec_fn: Any,
    ) -> None:
        result = exec_fn(f"git checkout -b {self._branch_name}")
        if result.exit_code != 0:
            raise RuntimeError(f"Failed to create branch: {result.stderr}")
        state.branch_created = True
        self._log.info(
            "[WRITE_BACK] execution=%s entry=%s — created branch %s",
            self._execution_id, entry_name, self._branch_name,
        )

    async def _commit_and_push(
        self,
        entry_name: str,
        state: _EntryState,
        exec_fn: Any,
        commit_msg: str,
    ) -> None:
        exec_fn("git add -A")

        commit_result = exec_fn(
            f'git commit -m "{commit_msg}"',
            timeout=30,
        )
        if commit_result.exit_code != 0:
            raise RuntimeError(f"Failed to commit: {commit_result.stderr}")

        state.commit_count += 1

        sha_result = exec_fn("git rev-parse HEAD")
        state.last_commit_sha = (sha_result.stdout or "").strip()

        if state.commit_count == 1:
            push_result = exec_fn(
                f"git push -u origin {self._branch_name}",
                timeout=60,
            )
        else:
            push_result = exec_fn("git push", timeout=60)

        if push_result.exit_code != 0:
            raise RuntimeError(f"Failed to push: {push_result.stderr}")

        self._log.info(
            "[WRITE_BACK] execution=%s entry=%s — commit #%d pushed "
            "(sha=%s)",
            self._execution_id,
            entry_name,
            state.commit_count,
            state.last_commit_sha[:12],
        )

    async def _create_pr(
        self,
        entry_name: str,
        state: _EntryState,
        eligible: _EligibleEntry,
        exec_fn: Any,
    ) -> None:
        meta = eligible.provision_result.git_metadata

        try:
            owner, repo = parse_github_repo(meta.repo_url)
        except ValueError as exc:
            raise RuntimeError(str(exc)) from exc

        if not state.github_token:
            cred_result = exec_fn(f"cat {_CREDENTIAL_FILE}", timeout=5)
            if cred_result.exit_code != 0:
                raise RuntimeError("Git credentials not available for PR creation")
            state.github_token = parse_token_from_credentials(cred_result.stdout)
            state.github_owner = owner
            state.github_repo = repo

        pr_title = f"Agent changes ({self._short_id})"
        pr_body = (
            f"Automated pull request from Stigmer agent execution.\n\n"
            f"**Execution:** `{self._execution_id}`\n"
            f"**Workspace:** `{entry_name}`\n"
        )

        try:
            pr_data = await github_create_pr(
                token=state.github_token,
                owner=state.github_owner,
                repo=state.github_repo,
                title=pr_title,
                body=pr_body,
                head=self._branch_name,
                base=eligible.base_branch,
            )
        except httpx.HTTPStatusError as exc:
            status_code = exc.response.status_code
            try:
                error_body = exc.response.json()
                message = error_body.get("message", str(exc))
            except Exception:
                message = str(exc)
            raise RuntimeError(
                f"GitHub API error (HTTP {status_code}): {message}"
            ) from exc
        except httpx.TimeoutException as exc:
            raise RuntimeError("GitHub API request timed out") from exc

        state.pr_created = True
        state.pr_url = pr_data.get("html_url", "")
        state.pr_number = pr_data.get("number", 0)

        self._log.info(
            "[WRITE_BACK] execution=%s entry=%s — PR #%d created: %s",
            self._execution_id,
            entry_name,
            state.pr_number,
            state.pr_url,
        )

    def _update_status(
        self,
        entry_name: str,
        state: _EntryState,
        eligible: _EligibleEntry,
        exec_fn: Any,
    ) -> None:
        """Build and publish a WorkspaceWriteBack proto to the status builder."""
        summary_result = exec_fn(
            f"git diff --stat {eligible.base_branch}...HEAD"
        )

        phase = (
            WorkspaceWriteBackPhase.WORKSPACE_WRITE_BACK_PR_CREATED
            if state.pr_created
            else WorkspaceWriteBackPhase.WORKSPACE_WRITE_BACK_PUSHED
        )

        wb = WorkspaceWriteBack(
            workspace_entry_name=entry_name,
            branch_name=self._branch_name,
            base_branch=eligible.base_branch,
            commit_sha=state.last_commit_sha,
            pull_request_url=state.pr_url,
            pull_request_number=state.pr_number,
            diff_summary=(summary_result.stdout or "").strip(),
            phase=phase,
        )
        self._sb.add_workspace_write_back(wb)

File: backend/services/runner/src/activities/execute-deep-agent/writeback-coordinator.ts
/**
 * Incremental git write-back coordinator for workspace entries.
 *
 * During agent execution, each file-modifying tool call triggers an
 * incremental commit-and-push cycle for the affected git workspace.
 * The first cycle creates the branch and PR; subsequent cycles add
 * commits to the same branch — the PR updates automatically on GitHub.
 *
 * DD-5: Incremental writeback, not batch.
 *
 * Lifecycle:
 *   1. Created after workspace provisioning in index.ts.
 *   2. `onFileModified(path)` called from streaming on each write/edit tool end.
 *   3. `finalize()` called from post-stream as a safety net.
 *
 * Concurrency: one mutex per workspace entry serializes git operations.
 */

import { create } from "@bufbuild/protobuf";
import {
  WorkspaceWriteBackSchema,
} from "@stigmer/protos/ai/stigmer/agentic/agentexecution/v1/writeback_pb";
import {
  WorkspaceWriteBackPhase,
} from "@stigmer/protos/ai/stigmer/agentic/agentexecution/v1/writeback_pb";
import { GitWriteBackMode } from "@stigmer/protos/ai/stigmer/agentic/session/v1/enum_pb";
import type { WorkspaceEntry } from "@stigmer/protos/ai/stigmer/agentic/session/v1/workspace_pb";
import type { WorkspaceBackend, ProvisionResult } from "../../shared/workspace/types.js";
import { SourceType } from "../../shared/workspace/types.js";
import type { StatusBuilder } from "./status-builder.js";

const WRITE_BACK_ENABLED_MODES = new Set([
  GitWriteBackMode.GIT_WRITE_BACK_MODE_UNSPECIFIED,
  GitWriteBackMode.GIT_WRITE_BACK_BRANCH_AND_PR,
]);

interface EntryState {
  branchCreated: boolean;
  prCreated: boolean;
  prUrl: string;
  prNumber: number;
  commitCount: number;
  lastCommitSha: string;
  githubToken: string;
  githubOwner: string;
  githubRepo: string;
}

interface EligibleEntry {
  readonly provisionResult: ProvisionResult;
  readonly baseBranch: string;
  readonly rootDir: string;
  readonly entryName: string;
}

export class WriteBackCoordinator {
  private readonly statusBuilder: StatusBuilder;
  private readonly executionId: string;
  private readonly workspaceBackend: WorkspaceBackend;
  private readonly shortId: string;
  private readonly branchName: string;

  private readonly eligible = new Map<string, EligibleEntry>();
  private readonly state = new Map<string, EntryState>();
  private readonly locks = new Map<string, Promise<void>>();

  constructor(opts: {
    statusBuilder: StatusBuilder;
    executionId: string;
    provisionResults: readonly ProvisionResult[];
    workspaceEntries: readonly WorkspaceEntry[];
    workspaceBackend: WorkspaceBackend;
  }) {
    this.statusBuilder = opts.statusBuilder;
    this.executionId = opts.executionId;
    this.workspaceBackend = opts.workspaceBackend;
    this.shortId = opts.executionId.slice(0, 8);
    this.branchName = `stigmer/${this.shortId}`;

    this.initEligibleEntries(opts.provisionResults, opts.workspaceEntries);
  }

  get hasEligibleEntries(): boolean {
    return this.eligible.size > 0;
  }

  /**
   * Called after a file-modifying tool completes. Resolves the path to
   * a workspace entry and runs an incremental commit/push cycle.
   * Fire-and-forget: errors are logged, never thrown.
   */
  async onFileModified(path: string): Promise<void> {
    try {
      const entryName = this.resolveEntry(path);
      if (!entryName) return;

      await this.withLock(entryName, () =>
        this.incrementalWriteBack(entryName),
      );
    } catch (err) {
      console.warn(
        `[WriteBack] execution=${this.executionId} — ` +
        `onFileModified error for '${path}': ${err}`,
      );
    }
  }

  /**
   * Post-execution safety net. Checks every eligible workspace entry
   * for remaining uncommitted changes and commits/pushes them.
   */
  async finalize(): Promise<void> {
    for (const entryName of this.eligible.keys()) {
      try {
        await this.withLock(entryName, () =>
          this.incrementalWriteBack(entryName),
        );
      } catch (err) {
        console.warn(
          `[WriteBack] execution=${this.executionId} entry=${entryName} — ` +
          `finalize error: ${err}`,
        );
      }
    }
  }

  // ── Initialization ──────────────────────────────────────────────────

  private initEligibleEntries(
    provisionResults: readonly ProvisionResult[],
    workspaceEntries: readonly WorkspaceEntry[],
  ): void {
    const modeMap = new Map<string, GitWriteBackMode>();
    for (const entry of workspaceEntries) {
      const source = entry.source;
      if (source?.source.case === "gitRepo") {
        modeMap.set(entry.name, source.source.value.writeBackMode);
      }
    }

    for (const pr of provisionResults) {
      if (pr.sourceType !== SourceType.GIT_REPO) continue;
      if (!pr.gitMetadata) continue;
      if (!pr.gitMetadata.gitCredentialsConfigured) continue;

      const mode = modeMap.get(pr.entryName) ?? GitWriteBackMode.GIT_WRITE_BACK_MODE_UNSPECIFIED;
      if (!WRITE_BACK_ENABLED_MODES.has(mode)) continue;

      this.eligible.set(pr.entryName, {
        provisionResult: pr,
        baseBranch: pr.gitMetadata.branch,
        rootDir: pr.rootDir,
        entryName: pr.entryName,
      });

      this.state.set(pr.entryName, {
        branchCreated: false,
        prCreated: false,
        prUrl: "",
        prNumber: 0,
        commitCount: 0,
        lastCommitSha: "",
        githubToken: "",
        githubOwner: "",
        githubRepo: "",
      });
    }

    if (this.eligible.size > 0) {
      console.log(
        `[WriteBack] execution=${this.executionId} — coordinator initialized with ` +
        `${this.eligible.size} eligible workspace(s): ${[...this.eligible.keys()].join(", ")}`,
      );
    }
  }

  // ── Path Resolution ─────────────────────────────────────────────────

  private resolveEntry(path: string): string | null {
    if (this.eligible.size === 0) return null;
    if (this.eligible.size === 1) return this.eligible.keys().next().value!;

    const normalized = path.replace(/^\/+/, "");
    for (const entryName of this.eligible.keys()) {
      if (normalized.startsWith(entryName + "/") || normalized === entryName) {
        return entryName;
      }
    }
    return null;
  }

  // ── Core Incremental Write-Back ─────────────────────────────────────

  private async incrementalWriteBack(entryName: string): Promise<void> {
    const entry = this.eligible.get(entryName)!;
    const entryState = this.state.get(entryName)!;
    const rootDir = entry.rootDir;

    const exec = async (cmd: string): Promise<string> => {
      return this.workspaceBackend.execute(`cd ${rootDir} && ${cmd}`);
    };

    let mutationStarted = false;
    try {
      const hasChanges = await this.hasChanges(exec);
      if (!hasChanges) return;

      mutationStarted = true;

      if (!entryState.branchCreated) {
        await this.createBranch(entryName, entryState, exec);
      }

      const commitMsg = `agent changes (${entryState.commitCount + 1})`;
      await this.commitAndPush(entryName, entryState, exec, commitMsg);

      if (!entryState.prCreated) {
        await this.createPr(entryName, entryState, entry, exec);
      }

      await this.updateStatus(entryName, entryState, entry, exec);

    } catch (err) {
      console.warn(
        `[WriteBack] execution=${this.executionId} entry=${entryName} — ` +
        `incremental error: ${err}`,
      );
      if (!mutationStarted) return;

      const wb = create(WorkspaceWriteBackSchema, {
        workspaceEntryName: entryName,
        baseBranch: entry.baseBranch,
        branchName: entryState.branchCreated ? this.branchName : "",
        phase: WorkspaceWriteBackPhase.WORKSPACE_WRITE_BACK_FAILED,
        error: String(err),
      });
      if (entryState.prCreated) {
        wb.pullRequestUrl = entryState.prUrl;
        wb.pullRequestNumber = entryState.prNumber;
      }
      this.statusBuilder.addWriteBack(wb);
    }
  }

  // ── Git Operations ──────────────────────────────────────────────────

  private async hasChanges(exec: (cmd: string) => Promise<string>): Promise<boolean> {
    const diff = await exec("git diff --stat").catch(() => "");
    const staged = await exec("git diff --cached --stat").catch(() => "");
    if (diff.trim() || staged.trim()) return true;

    const untracked = await exec("git ls-files --others --exclude-standard").catch(() => "");
    return untracked.trim().length > 0;
  }

  private async createBranch(
    entryName: string,
    entryState: EntryState,
    exec: (cmd: string) => Promise<string>,
  ): Promise<void> {
    await exec(`git checkout -b ${this.branchName}`);
    entryState.branchCreated = true;
    console.log(
      `[WriteBack] execution=${this.executionId} entry=${entryName} — ` +
      `created branch ${this.branchName}`,
    );
  }

  private async commitAndPush(
    entryName: string,
    entryState: EntryState,
    exec: (cmd: string) => Promise<string>,
    commitMsg: string,
  ): Promise<void> {
    await exec("git add -A");
    await exec(`git commit -m "${commitMsg}"`);

    entryState.commitCount++;
    const shaOutput = await exec("git rev-parse HEAD");
    entryState.lastCommitSha = shaOutput.trim();

    if (entryState.commitCount === 1) {
      await exec(`git push -u origin ${this.branchName}`);
    } else {
      await exec("git push");
    }

    console.log(
      `[WriteBack] execution=${this.executionId} entry=${entryName} — ` +
      `commit #${entryState.commitCount} pushed (sha=${entryState.lastCommitSha.slice(0, 12)})`,
    );
  }

  private async createPr(
    entryName: string,
    entryState: EntryState,
    entry: EligibleEntry,
    _exec: (cmd: string) => Promise<string>,
  ): Promise<void> {
    const meta = entry.provisionResult.gitMetadata!;
    const { owner, repo } = parseGithubRepo(meta.repoUrl);

    if (!entryState.githubToken) {
      entryState.githubToken = extractGithubToken(meta.repoUrl);
      entryState.githubOwner = owner;
      entryState.githubRepo = repo;
    }

    const prTitle = `Agent changes (${this.shortId})`;
    const prBody =
      `Automated pull request from Stigmer agent execution.\n\n` +
      `**Execution:** \`${this.executionId}\`\n` +
      `**Workspace:** \`${entryName}\`\n`;

    const resp = await fetch(
      `https://api.github.com/repos/${entryState.githubOwner}/${entryState.githubRepo}/pulls`,
      {
        method: "POST",
        headers: {
          "Authorization": `Bearer ${entryState.githubToken}`,
          "Accept": "application/vnd.github+json",
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          title: prTitle,
          body: prBody,
          head: this.branchName,
          base: entry.baseBranch,
        }),
      },
    );

    if (!resp.ok) {
      const body = await resp.text();
      throw new Error(`GitHub API error (HTTP ${resp.status}): ${body}`);
    }

    const data = await resp.json() as { html_url?: string; number?: number };
    entryState.prCreated = true;
    entryState.prUrl = data.html_url ?? "";
    entryState.prNumber = data.number ?? 0;

    console.log(
      `[WriteBack] execution=${this.executionId} entry=${entryName} — ` +
      `PR #${entryState.prNumber} created: ${entryState.prUrl}`,
    );
  }

  private async updateStatus(
    entryName: string,
    entryState: EntryState,
    entry: EligibleEntry,
    exec: (cmd: string) => Promise<string>,
  ): Promise<void> {
    const summaryOutput = await exec(
      `git diff --stat ${entry.baseBranch}...HEAD`,
    ).catch(() => "");

    const phase = entryState.prCreated
      ? WorkspaceWriteBackPhase.WORKSPACE_WRITE_BACK_PR_CREATED
      : WorkspaceWriteBackPhase.WORKSPACE_WRITE_BACK_PUSHED;

    const wb = create(WorkspaceWriteBackSchema, {
      workspaceEntryName: entryName,
      branchName: this.branchName,
      baseBranch: entry.baseBranch,
      commitSha: entryState.lastCommitSha,
      pullRequestUrl: entryState.prUrl,
      pullRequestNumber: entryState.prNumber,
      diffSummary: summaryOutput.trim(),
      phase,
    });

    this.statusBuilder.addWriteBack(wb);
  }

  // ── Concurrency ─────────────────────────────────────────────────────

  private async withLock(
    entryName: string,
    fn: () => Promise<void>,
  ): Promise<void> {
    const existing = this.locks.get(entryName) ?? Promise.resolve();
    const next = existing.then(fn, fn);
    this.locks.set(entryName, next);
    await next;
  }
}

// ── Helpers ───────────────────────────────────────────────────────────

/**
 * Extract GitHub owner and repo from a clone URL.
 * Supports HTTPS (with or without .git) and SSH formats.
 */
export function parseGithubRepo(repoUrl: string): { owner: string; repo: string } {
  const httpsMatch = repoUrl.match(
    /github\.com[/:]([^/]+)\/([^/.]+?)(?:\.git)?$/,
  );
  if (httpsMatch) {
    return { owner: httpsMatch[1], repo: httpsMatch[2] };
  }
  throw new Error(`Cannot parse GitHub owner/repo from URL: ${repoUrl}`);
}

/**
 * Extract a GitHub token from an HTTPS clone URL that has credentials embedded.
 * Format: https://{token}@github.com/...
 */
export function extractGithubToken(repoUrl: string): string {
  const match = repoUrl.match(/https?:\/\/([^@]+)@github\.com/);
  if (match) return match[1];

  const envToken = process.env.GITHUB_TOKEN;
  if (envToken) return envToken;

  throw new Error(
    "Cannot extract GitHub token from repo URL and GITHUB_TOKEN is not set",
  );
}
